draft-irtf-cfrg-fiat-shamir.md

```diff
@@ -107,59 +113,128 @@ This will be expanded in future versions of this specification.
+            response = self.sigma_protocol.deserialize_response(response_bytes)
+            commitment = self.sigma_protocol.simulate_commitment(response, challenge)
+
+            return self.sigma_protocol.verifier(commitment, challenge, response)
+
+        def prove_batchable(self, witness, rng):
+            # Proving method using commitment-response format.
+            # Allows for batching.
+            (commitment, challenge, response) = self._prove(witness, rng)
+            # running the verifier here is just a sanity check
+            assert self.sigma_protocol.verifier(commitment, challenge, response)
+            domain_sep = "NISigmaProtocolBatchableProof".encode('utf-8')
+            return domain_sep + self.sigma_protocol.serialize_commitment(commitment) + self.sigma_protocol.serialize_response(response)
+
+        def verify_batchable(self, proof):
+            # Before running the sigma protocol verifier, one must also check that:
+            # - serialized proof domain separator matches expected value
+            domain_sep = "NISigmaProtocolBatchableProof".encode('utf-8')
+            domain_sep_bytes = proof[:len(domain_sep)]
+            assert domain_sep_bytes == domain_sep
+
+            # - the proof length is exactly len(domain_sep) + commit_bytes_len + response_bytes_len
+            assert len(proof) == len(domain_sep) + self.sigma_protocol.instance.commit_bytes_len + self.sigma_protocol.instance.response_bytes_len
 
-## Codec for Linear maps {#group-prove}
+            # - proof deserialization successfully produces a valid commitment and a valid response
+            commitment_bytes = proof[len(domain_sep):self.sigma_protocol.instance.commit_bytes_len]
+            response_bytes = proof[len(domain_sep) + self.sigma_protocol.instance.commit_bytes_len:]
+            commitment = self.sigma_protocol.deserialize_commitment(commitment_bytes)
+            response = self.sigma_protocol.deserialize_response(response_bytes)
 
-We describe a codec for Schnorr proofs over groups of prime order `p` that is intended for duplex sponges where `Unit = u8`.
+            self.codec.prover_message(self.hash_state, commitment)
+            challenge = self.codec.verifier_challenge(self.hash_state)
+            return self.sigma_protocol.verifier(commitment, challenge, response)
 
-    class LinearMapCodec:
-        Group: groups.Group = None
-        DuplexSponge: DuplexSpongeInterface = None
+## NIZK instances (ciphersuites)
 
-        def init(self, iv: bytes):
-            self.hash_state = self.DuplexSponge(iv)
+We describe NIZK instances for combinations of protocols (SigmaProtocol), codec (Codec), and hash fuction (DuplexSpongeInterface). Descriptions of codecs and hash functions are in the following sections.
+
+    class NISchnorrProofShake128P256(NIZK):
+        Protocol = SchnorrProof
+        Codec = P256Codec
+        Hash = SHAKE128
+
+    class NISchnorrProofShake128Bls12381(NIZK):
```

Comment:
Does this PR now close #17?

Reply:
Not totally, we'd also want to define the ciphersuite in the spec group definitions themselves.

Reply:
Here: https://github.com/mmaker/draft-irtf-cfrg-sigma-protocols/blob/3ab5b8787ef9d0c9677a9afcd5e5f6613a7b45a5/poc/fiat_shamir.sage#L7

Reply:
Note that this code is outdated, I've now named all references to "NIZK" to be "NISigmaProtocol" instead, for specificity. So you shouldn't see NIZK referenced anywhere after this PR cleans it up.